crates/accelerate/src/gate_direction.rs

```diff
@@ -142,8 +159,471 @@ where
+
+    for (node, replacemanet_dag) in nodes_to_replace {
+        dag.py_substitute_node_with_dag(
+            py,
+            dag.get_node(py, node)?.bind(py),
+            &replacemanet_dag,
+            None,
+            true,
+        )?;
+    }
+
+    Ok(dag)
+}
+
+// Check whether the dag as calibration for a DAGOpNode
+fn has_calibration_for_op_node(
+    py: Python,
+    dag: &DAGCircuit,
+    packed_inst: &PackedInstruction,
+    qargs: &[Qubit],
+) -> PyResult<bool> {
+    let py_args = PyTuple::new_bound(py, dag.qubits().map_indices(qargs));
+
+    let dag_op_node = Py::new(
+        py,
+        (
+            DAGOpNode {
+                instruction: CircuitInstruction {
+                    operation: packed_inst.op.clone(),
+                    qubits: py_args.unbind(),
+                    clbits: PyTuple::empty_bound(py).unbind(),
+                    params: packed_inst.params_view().iter().cloned().collect(),
+                    extra_attrs: packed_inst.extra_attrs.clone(),
+                    #[cfg(feature = "cache_pygates")]
+                    py_op: packed_inst.py_op.clone(),
+                },
+                sort_key: "".into_py(py),
+            },
+            DAGNode { node: None },
+        ),
+    )?;
+
+    dag.has_calibration_for(py, dag_op_node.borrow(py))
+}
+
+// Return a replacement DAG for the given standard gate in the supported list
+// TODO: optimize it by caching the DAGs of the non-parametric gates and caching and
+// mutating upon request the DAGs of the parametric gates
+fn replace_dag(
+    py: Python,
+    std_gate: StandardGate,
+    inst: &PackedInstruction,
+) -> PyResult<DAGCircuit> {
+    let replacement_dag = match std_gate {
+        StandardGate::CXGate => cx_replacement_dag(py),
+        StandardGate::ECRGate => ecr_replacement_dag(py),
+        StandardGate::CZGate => cz_replacement_dag(py),
+        StandardGate::SwapGate => swap_replacement_dag(py),
+        StandardGate::RXXGate => rxx_replacement_dag(py, inst.params_view()),
+        StandardGate::RYYGate => ryy_replacement_dag(py, inst.params_view()),
+        StandardGate::RZZGate => rzz_replacement_dag(py, inst.params_view()),
+        StandardGate::RZXGate => rzx_replacement_dag(py, inst.params_view()),
+        _ => panic!("Mismatch in supported gates assumption"),
+    };
+
+    replacement_dag
+}
+
+//###################################################
+// Utility functions to build the replacement dags
+//
+// TODO: replace this once we have a Rust version of QuantumRegister
+#[inline]
+fn add_qreg(py: Python, dag: &mut DAGCircuit, num_qubits: u32) -> PyResult<Vec<Qubit>> {
+    let qreg = imports::QUANTUM_REGISTER
+        .get_bound(py)
+        .call1((num_qubits,))?;
+    dag.add_qreg(py, &qreg)?;
+    let mut qargs = Vec::new();
+
+    for i in 0..num_qubits {
+        let qubit = qreg.call_method1(intern!(py, "__getitem__"), (i,))?;
+        qargs.push(dag.qubits().find(&qubit).expect("Qubit should stored"));
```

Comment:
```suggestion
        qargs.push(dag.qubits().find(&qubit).expect("Qubit should have been stored in the DAGCircuit"));
```

Reply:
Ah... good catch. Thx. Done in https://github.com/Qiskit/qiskit/pull/13069/commits/79a62174af9c6955a3d9b05adcb456d939641950